### DISCUSSION TITLE
ty.cast not casting type in __new__ method when ty.Protocol is passed as a TypeVar to the class.

### DISCUSSION THREAD

**Discussion by CaffeineDuck** [microsoft/pyright#7465] [Q&A]
I'm implementing a Dependency Injection library for learning. This is my injectable class:

```python
class Injectable[T]:
    __injection__: type

    def __new__(cls, injection_class: type[T]) -> T:
        if injection_class == Injectable:
            raise ValueError("Injectable cannot be injected.")

        self = super().__new__(cls)
        self.__injection__ = injection_class
        return ty.cast(T, self)

    def __getattr__(self, _: str) -> None:
        raise DIConfigurationError()
```

This is example usage:

```python
from abc import ABC, abstractmethod

from injkt import Bind, Config, Injectable, Injktor, inject_args_deps


class IMailService(ABC):
    @abstractmethod
    def send_mail(self, subject: str, to: str) -> None:
        ...


class SmtpMailService(IMailService):
    def send_mail(self, subject: str, to: str) -> None:
        raise NotImplementedError()


injktor = Injktor(
    Config(
        {
            Bind(IMailService, SmtpMailService),
        }
    )
)


@inject_args_deps
def business_logic(mail_service=Injectable(IMailService)):
    mail_service.send_mail("Hello", "world")


business_logic()
```

The type of **mail_service** in the above function is `IMailService`, however if I use `ty.Protocol` instead of `ABC`:

```python
class IMailService(ty.Protocol):
    def send_mail(self, subject: str, to: str) -> None:
        ...
```

The type of **mail_service** in the **business_logic** function changes to `Injectable[IMailService]` which breaks the typings I'm trying to achieve.

I know changing the implementation detail will solve the problem, I'm not trying to figure out other ways for solving the problem, I'm just trying to figure out why this is happening and if this is solvable using the current implementation for DI.

For more reference, this is the github repo: <https://github.com/caffeineduck/injkt>

**Comment by erictraut** [comment]
It took me a long time to understand your report. For future bug reports and questions, please post a minimal, self-contained code sample that demonstrates the problem. Your code above is in multiple parts, refers to symbols in external libraries, and contains a bunch of extraneous lines that have nothing to do with the problem at hand.

Here's a minimal piece of code that (I think) gets to the heart of the issue you're seeing.

```python
from typing import Generic, Protocol, TypeVar
from abc import ABC

T = TypeVar("T")

class Injectable(Generic[T]):
    def __new__(cls, i: type[T]) -> T:
        return i()

class IMailService_Proto(Protocol): ...

class IMailService_ABC(ABC): ...

m1 = Injectable(IMailService_Proto)  # Mypy produces an error here, as it should
reveal_type(m1)

m2 = Injectable(IMailService_ABC)
reveal_type(m2)
```

According to the typing spec, a protocol class is not type compatible with `type[T]`, so the expression `Injectable(IMailService_Proto)` should produce a type error. Pyright should produce an error here, so this is a bug. I'll need to look into this in more detail to determine what's happening. This explains the behavior you're seeing when you use a protocol class versus a non-protocol (ABC) class.

You currently define `Injectable` as a generic class. It may be better to model it as a non-generic class with a generic `__new__` method. It doesn't appear that you use `T` anywhere else in the class.
